src/AccessControl/tests/test_formatter.py

```diff
@@ -36,14 +92,129 @@ def test_positional_argument_regression(self):
+                         '<Item public1>')
+        self.assertEqual(SafeFormatter(u'{0.public2}').safe_format(folder),
+                         '<Item public2>')
+        # But not the private item:
+        self.assertRaises(Unauthorized,
+                          SafeFormatter(u'{0.private}').safe_format,
+                          folder)
+
+    def test_prevents_bad_string_formatting_item(self):
+        from AccessControl.safe_formatter import SafeFormatter
+        # Accessing basic Python types in a basic Python dict is fine.
+        foo = {'__private__': 'not really private'}
+        self.assertEqual(SafeFormatter('{0[__private__]}').safe_format(foo),
+                         'not really private')
+        # But for non-basic items or non-basic lists, we want run checks.
+        folder = self._create_folder_with_mixed_contents()
+        # We can get the public items just fine:
+        self.assertEqual(SafeFormatter('{0[public1]}').safe_format(folder),
+                         '<Item public1>')
+        self.assertEqual(SafeFormatter('{0[public2]}').safe_format(folder),
+                         '<Item public2>')
+        # But not the private item:
+        self.assertRaises(Unauthorized,
+                          SafeFormatter('{0[private]}').safe_format,
+                          folder)
+
+    def test_prevents_bad_unicode_formatting_item(self):
+        from AccessControl.safe_formatter import SafeFormatter
+        # Accessing basic Python types in a basic Python dict is fine.
+        foo = {'__private__': 'not really private'}
+        self.assertEqual(SafeFormatter(u'{0[__private__]}').safe_format(foo),
+                         'not really private')
+        # But for non-basic items or non-basic lists, we want run checks.
+        folder = self._create_folder_with_mixed_contents()
+        # We can get the public items just fine:
+        self.assertEqual(SafeFormatter(u'{0[public1]}').safe_format(folder),
+                         '<Item public1>')
+        self.assertEqual(SafeFormatter(u'{0[public2]}').safe_format(folder),
+                         '<Item public2>')
+        # But not the private item:
+        self.assertRaises(Unauthorized,
+                          SafeFormatter(u'{0[private]}').safe_format,
+                          folder)
+
+    def test_prevents_bad_string_formatting_key(self):
+        from AccessControl.safe_formatter import SafeFormatter
+        from AccessControl.ZopeGuards import guarded_getitem
+        from persistent.list import PersistentList
+        # Accessing basic Python types in a basic Python list is fine.
+        foo = list(['__private__'])
+        self.assertEqual(SafeFormatter('{0[0]}').safe_format(foo),
+                         '__private__')
+        self.assertEqual(guarded_getitem(foo, 0), '__private__')
+        # For basic Python types in a non-basic list, we guard the access.
+        foo = PersistentList(foo)
```

Comment:
I would prefer that this test *not* use strings like `'__private__'` in this example:  the point of `__allow_access_to_unprotected_subobjects` is that it denies access *regardless* of the contents for items without explicit protection.

Reply:
Fair point. I have changed the test.